File: 1036-RottingOranges/1036-RottingOranges.cpp
// Last updated: 17/07/2025, 20:51:42
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        int count = 0;
        int level = 0; 
        // find the level from different rotten oranges 
        queue<pair<pair<int,int>,int>> q;
        int n = grid.size();
        int m = grid[0].size();
        // int vis[n][m];
        vector<vector<int>> vis(n, vector<int>(m, 0));
        int fresh = 0;
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(grid[i][j] == 1){
                    fresh ++;
                }
                if(grid[i][j] == 2){
                    q.push({{i,j},0});
                    vis[i][j] = 2;
                }else{
                    vis[i][j] = 0;
                }
            }
        }
        int maxTime = 0;
        int rotten = 0;
        while(!q.empty()){
            int row = q.front().first.first;
            int col = q.front().first.second;
            int time = q.front().second;
            q.pop();
            int drow[4] = {-1,0,1,0};
            int dcol[4] = {0,1,0,-1};
            maxTime = max(maxTime, time);
            for(int i = 0; i< 4; i++){
                int nrow = row+drow[i];
                int ncol = col + dcol[i];
                if(nrow >=0 && nrow<n && ncol>= 0 && ncol < m
                 && vis[nrow][ncol] != 2 && grid[nrow][ncol] == 1){
                    rotten ++;
                    q.push({{nrow, ncol}, time+1});
                    vis[nrow][ncol] = 2;
                }
            }
        }
        if(rotten != fresh){
            return -1;
        }
        return maxTime;
        
    }
};

File: 1036-RottingOranges/1036-RottingOranges.py
# Last updated: 01/10/2025, 00:12:24
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows,cols = len(grid), len(grid[0])

        q = deque()
        fresh = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 2:
                    q.append((r,c,0))
                elif grid[r][c] == 1:
                    fresh +=1
        
        time = 0
        while q:
            r,c,t = q.popleft()
            time = max(time,t)
            for dr,dc in [(1,0),(-1,0), (0,1), (0,-1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < rows and 0<=nc < cols and grid[nr][nc] == 1:
                    grid[nr][nc] = 2
                    fresh -= 1
                    q.append((nr,nc,t+1))
        
        return time if fresh == 0 else -1